New basic Request/Response transformers

mskacelik opened the pull request:
- New transformers:
    - Header modifiers (request/response)
    - Body modifiers/changers (request/response) => very basic API (no `JSON`/`XML`/... body support/handling, only `PLAIN_TEXT` (buffer))
- refactored/migrated tests
- updated `.gitignore`

I was also planning on adding `QueryParam` transformers and some `Guard` (Fault Tolerance) transformers,  but they are a bit more complex and need some discussion.

tsegismont commented:
@mskacelik have you checked out upstream transformers ? https://github.com/eclipse-vertx/vertx-http-proxy/tree/main/src/main/java/io/vertx/httpproxy

They are not available in Vert.x 4, but you could inspire from the work done there.

For example, when you want to modify a body, replacing it with a fixed one is useful in a limited set of use cases. Sometimes users want to buffer the backend body, modify it (e.g. removing some JSON fields) and then reply.

jponge reviewed (commented):
Nice work, and agreed with @tsegismont 👍 

It seems like the request modifiers do take a function, but the tests focus on swapping a body with another, so we should add a JSON object transformation test.

I'm not sure we need a higher-level and content-aware body transformer _at this stage_, just taking a body, buffering it and converting to JSON, then making some in-place transformations should be our reasonable target.

jponge commented:
> I have just added the JSON body transformer, given that it is by far the most complex transformer that we have (due to stream handling), I would really appreciate it if you could check the logic. Plus some notes:

The streaming part looks good to me.

> * is upon JSON de-serialization failure returning failed future good idea? (in catch block), this will be returned by proxy as 502 response.

Yes, a failed `Future` is how you report asynchronous failures. I wonder if we should not use a `400` status code here, because if we have JSON parsing exception, then it's a bad request from the client, not a gateway problem.

> * I also implemented the implicit content type change upon usage of `(Response/Request)JSONBodyModifier(JsonObject)`, but maybe its not the best and most straightforward solution, instead of having the user handle this by themself (via `Reponse/RequestContentTypeModifier`).

From what I see in tests, it's a sensible API. What alternatives did you have in mind?

> Also added simple transformers for HTTP Method and Content Type changing.

👍 
 
> Regarding the limit, I happen to notice that the [BodyTransformers](https://github.com/eclipse-vertx/vertx-http-proxy/blob/main/src/main/java/io/vertx/httpproxy/BodyTransformers.java) in 5.x, are handling the body buffer limit as part of the original API – with default `DEFAULT_MAX_BUFFERED_SIZE := 256 * 1024`, so maybe we can also do it like that, instead of creating a separate API for "safe" variation that enforces the limit, WDYT?

So we _could_ have a body limit enforced when we make transformations, but maybe we will also want to have transformers that just check the body limit. I'm not against having a limit parameter here, my real concern is some duplication 😉  (@tsegismont WDYT?)

mskacelik commented:
Hopefully, the last feature change for this PR.
- removed the "safe" content-type change to `JSON`, I think it's better for the user to change it if it's really needed.
- added ProxyResponseFactory for better handling of failed Futures + added tests to check if it breaks the chain of transformers (both for response and request)
- added the content length limit guards transformers (request/response)

tsegismont reviewed (commented):
Overall, it looks good to me.

Looking at the json modifier specifically, keep in mind that the current proposal will only allow to transform from JSON object to JSON object, excluding other transformations like JSON array to JSON array, JSON array to JSON object... etc

jponge commented:
If you think it's easy, go for it!

mskacelik commented:
In theory, we could even create very specific transformers for the transformation of `Function<JsonObject, JsonArray>` and `Function<JsonArray, JsonObject>` individually. I don't think these implementations would be problematic since I already started with some abstraction.

But again, is it such an important use case to bloat the project with such specific use case scenarios?

mskacelik commented:
Added:
 - JsonArray transformers
 - JsonArray -> JsonObject transformers
 - JsonObject -> JsonArray transformers

And quite a hefty refactoring of the Abstract Json classes.
(made it a separate commit)

mskacelik commented:
> In Vert.x 5 we have transformers for Object->Object:
> 
> https://github.com/eclipse-vertx/vertx-http-proxy/blob/af134bbb2da9b1ffb4d311ea650a6b1a66f99eb2/src/main/java/io/vertx/httpproxy/BodyTransformers.java#L87-L97
> 
> The Vert.x JSON API allows to decode an arbitrary payload. Then, users can inspect the type return (object, array, string, boolean, null...)

So in (Response/Request)BodyModifier instead of having a Body parameter, we would have something like `BiFunction<ProxyContext, Object, Object>`,`Function<Object, Object>`, and `Object`?
We would use the `Json` encoder to get the `Object`'s representation as `JsonValue`. And since we are not implicitly changing the content type, there is no need to handle explicitly the text and numbers as plain text.

And I guess there is no need for `Object -> JsonArray`, `JsonObject -> Object` ... (and other permutations).

jponge commented:
@tsegismont WDYT?
